feat: Add local error injection demo with KIND cluster

ksaur opened the pull request:
## Summary

Adds a new local demonstration of NVSentinel's GPU fault detection and auto-quarantine capabilities using KIND (Kubernetes in Docker). The demo runs entirely on a laptop without requiring real GPU hardware or cloud infrastructure.

The goal of this feature is to enable users to quickly test out NVSentinel without setting up cloud/infra/GPU, and just learn about the features with a quick local demo.


**Local XID Demo** (`demos/local-xid-demo/`):
- Complete end-to-end demo showing XID 79 error detection and node cordoning
- Uses fake DCGM to simulate GPU failures
- Demonstrates the full event flow: DCGM → GPU Health Monitor → Platform Connectors → MongoDB → FQM → Node Cordon

**Demo includes:**
- Automated setup script (KIND cluster + minimal NVSentinel deployment)
- XID error injection using `dcgmi test --inject`
- Verification script with retry logic
- Cleanup script
- Comprehensive README with screenshots and troubleshooting

## Type of Change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change
- [x] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [x] Documentation/CI
- [ ] Fault Management
- [ ] Health Monitors
- [ ] Janitor
- [x] Other: Demo

## Testing
- [ ] Tests pass locally
- [x] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [x] Documentation updated (if needed)
- [x] Ready for review





<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive local fault-injection demo documentation with prerequisites, setup instructions, phase-by-phase walkthroughs, troubleshooting guidance, and configuration details.

* **New Features**
  * Added automated demo workflow enabling users to run a complete fault detection and node quarantine demonstration on a local KIND cluster with a single command.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lalitadithya commented on demos/local-xid-demo/README.md:
````diff
@@ -0,0 +1,372 @@
+
+```bash
+# Use a specific version
+NVSENTINEL_VERSION=v0.4.0 ./scripts/00-setup.sh
````
```suggestion
NVSENTINEL_VERSION=v0.3.0 ./scripts/00-setup.sh
```

lalitadithya commented on demos/local-xid-demo/scripts/02-inject-xid.sh:
```diff
@@ -0,0 +1,138 @@
+    log "Found DCGM pod: $dcgm_pod"
+    log "Injecting XID 79 error via DCGM..."
+    
+    # Inject field 84 (Inforom Valid) = 0, which triggers XID 79 error
```
yes, much better. Thank you!

ksaur committed:
Add local XID demo for NVSentinel testing

Add a complete local demo environment for testing NVSentinel without GPU
hardware. The demo runs in a KIND cluster with simulated GPU failures.

Features:
- Automated setup with Makefile (make demo for full workflow)
- Step-by-step scripts for manual execution
- Fake DCGM for GPU fault simulation
- Single-member MongoDB replica set for minimal resources
- Complete end-to-end testing of fault detection and node quarantine

Demo components:
- 00-setup.sh: Creates cluster, installs NVSentinel with optimized config
- 01-show-cluster.sh: Displays cluster status and health
- 02-inject-xid.sh: Simulates GPU hardware fault via DCGM injection
- 03-verify-cordon.sh: Verifies automated node cordoning with retry logic
- 99-cleanup.sh: Removes demo cluster

Improvements:
- Fixed DCGM race condition with readiness probe
- Optimized MongoDB to use single replica (faster startup)
- Added proper error handling for CRD installation and pod readiness
- Clarified GPU fault injection terminology
- Added demo section to root README

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
Fix pod label selectors in wait_for_pods

The label selectors were incorrect, causing kubectl wait to not find
the pods and timeout. Updated to use the correct Kubernetes standard
labels:
- Platform Connectors: app.kubernetes.io/name=nvsentinel
- Fault Quarantine: app.kubernetes.io/name=fault-quarantine

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
simplify output

Signed-off-by: Karla Saur <ksaur@nvidia.com>

lalitadithya commented on demos/local-xid-demo/data/xid-79-error.json:
I don't think we are using this json, should we drop it?

lalitadithya reviewed (commented):
overall looks good, just some nits

ksaur committed:
Merge branch 'main' into ksaur/demo-xid

ksaur committed:
Address code review feedback

- Rename 02-inject-xid.sh to 02-inject-error.sh for accuracy
- Remove unused xid-79-error.json file
- Update all references in Makefile, README, and scripts

ksaur committed:
Add jq to prerequisites check

- Add jq to prerequisite checks in 00-setup.sh
- Update README to list jq as required (not optional)
- Remove trailing whitespace from 99-cleanup.sh

ksaur committed:
Rename demo and generalize terminology

- Rename demos/local-xid-demo to demos/local-fault-injection-demo
- Update all references in README files
- Change 'XID Error' to 'GPU Error' in section headers and messages
- Update function names (inject_xid_error -> inject_gpu_error)
- Make demo description more generic and less XID-focused
- Keep XID references where technically accurate (e.g., detection methods)
